### DISCUSSION TITLE
How to call a method which expects a pointer?

### DISCUSSION THREAD

**Discussion by stiezinger** [microsoft/CsWin32#250] [Q&A] [answered]
I want to call the method DhcpEnumSubnets() from the Dhcpsapi.dll. 
CsWin32 autogenerated the method and the structs needed to call the method. 
Here is my example:

```
public class Win32Example
{
    public static void EnumSubnets()
    {
        uint resumeHandle = 0;
        uint preferredMaximum = 65536;
        DHCP_IP_ARRAY* enumInfo = new DHCP_IP_ARRAY*();
        var result = PInvoke.DhcpEnumSubnets("10.130.224.11", 
                                            ref resumeHandle, 
                                            preferredMaximum, 
                                            ref enumInfo, 
                                            out var elementsRead, 
                                            out var elementsTotal);
    }
}
```

The variable enumInfo should have the desired values after a succesfull call. 
It has to be passed as ref, but for that I have to initialize it. 
I receive the error:

CS0214	Pointers and fixed size buffers may only be used in an unsafe context

After marking the class Win32Example with the unsafe keyword:

```
public unsafe class Win32Example
{
...
}
```
The error changes to:

CS1919	Unsafe type 'DHCP_IP_ARRAY*' cannot be used in object creation

I discovered the keyword stackalloc, and changed initialization of enumInfo to:

```
...
DHCP_IP_ARRAY* enumInfo = stackalloc DHCP_IP_ARRAY[0];
...
```

Now I receive an alert stating that all threads were allowed to run to prevent an unsafe abort when evaluating the method.
My goal is to call the method from a controller in a asp.net core web api and pass the subnets as the result to the request.

The method call looks now like this:

```
public unsafe class Win32Example
{
    public static void EnumSubnets()
    {
        uint resumeHandle = 0;
        uint preferredMaximum = 65536;
        DHCP_IP_ARRAY* enumInfo = stackalloc DHCP_IP_ARRAY[0];
        var result = PInvoke.DhcpEnumSubnets("10.130.224.11", 
                                            ref resumeHandle, 
                                            preferredMaximum, 
                                            ref enumInfo, 
                                            out var elementsRead, 
                                            out var elementsTotal);
    }
}
```

Is this the right way to call this method? If this is the right way, how to I marshal the DHCP_IP_ARRAY* to an uint array, 
to get the actual subnet addresses? Marshal.PtrToStructure() from the System.Runtime.InteropServices requires a IntPtr.
The structure DHCP_IP_ADDRESS for an IPv4 address, is currently missing, maybe because it is just an uint.

________________________________________________________________________________________________________________________

Thank you AArnott for your help. For further reference, this worked for me:

```csharp 
public static unsafe List<string> DhcpEnumSubnetAddresses(string serverAddress)
{
    uint resumeHandle = 0;
    uint preferredMaximum = 65536;
    DHCP_IP_ARRAY* enumInfo = null;

    List<string> subnets = new();

    var result = PInvoke.DhcpEnumSubnets(serverAddress,
        ref resumeHandle,
        preferredMaximum,
        ref enumInfo,
        out uint elementsRead,
        out uint elementsTotal);

    try
    {
        if (result.IsEqual(DhcpErrorCode.ERROR_NO_MORE_ITEMS))
            return subnets;

        if (result.IsSuccess())
        {
            DHCP_IP_ARRAY subnetElements = *enumInfo;
            for (uint j = 0; j < subnetElements.NumElements; j++)
            {
                uint subnetAddressNative = subnetElements.Elements[j];
                var subnetAddress = DhcpHelper.NativeIpAddressToString(subnetAddressNative);
                subnets.Add(subnetAddress);
            }
        }
    }
    finally
    {
        PInvoke.DhcpRpcFreeMemory(enumInfo);
    }
    return subnets;
}
```
Notice the `DHCP_IP_ARRAY subnetElements = *enumInfo` part.
I am doing this to avoid accessing the first index of `enumInfo`,
there are all the desired values.

`DhcpHelper.NativeIpAddressToString()` is just a helper method to 
convert the uint ipv4 address into the string representation.

`IsEqual()` and `IsSuccess()` are extension methods to compare the
uint `result` value with the value from the enum DhcpErrorCode.

**Comment by AArnott** [comment, marked as answer]
Thanks for reaching out. This is a sub-par experience and I filed https://github.com/microsoft/win32metadata/issues/452 to get this fixed in the metadata. The only real problem is the `EnumInfo` parameter should be `out` instead of `ref`. So just set it to `null`.

In the meantime, this syntax should work. You don't want to use `Marshal.PtrToStructure` probably because you need the original pointer to release the memory, and honestly the syntax doesn't get any easier anyway.

```cs
using System;
using Microsoft.Windows.Sdk;

unsafe
{
    uint resumeHandle = 0;
    uint preferredMaximum = 65536;
    DHCP_IP_ARRAY* enumInfo = null;
    while (PInvoke.DhcpEnumSubnets("192.168.1.1",
        ref resumeHandle,
        preferredMaximum,
        ref enumInfo,
        out uint elementsRead,
        out uint elementsTotal) == Constants.ERROR_SUCCESS)
    {
        try
        {
            for (uint i = 0; i < elementsRead; i++)
            {
                DHCP_IP_ARRAY info = enumInfo[i];
                for (uint j = 0; j < info.NumElements; j++)
                {
                    uint ipv4 = info.Elements[j];
                }
            }
        }
        finally
        {
            for (uint i = 0; i < elementsRead; i++)
            {
                PInvoke.DhcpRpcFreeMemory(enumInfo[i].Elements);
            }

            PInvoke.DhcpRpcFreeMemory(enumInfo);
        }
    }
}
```

And don't forget to [release the memory](https://docs.microsoft.com/en-us/windows/win32/api/dhcpsapi/nf-dhcpsapi-dhcpenumsubnets#remarks) when you're done with it.

> The structure DHCP_IP_ADDRESS for an IPv4 address, is currently missing, maybe because it is just an uint.

Ya, that's probably why they did this in the metadata.

**Reply by stiezinger** [reply]
Thank you very much, this helped me.

**Reply by stiezinger** [reply]
Hello AArnott,

I am sorry to bother you, may I ask one more question? I have tried your solution to receive the subnet addresses. My dhcp server has 4 scopes. This is indeed indicated in `elementsRead`, it has the value 4. The outer for loop works without problems, the inner for loop is causing some weird happenings, in the first iteration the `info.NumElements` field has the value 4. In the second iteration `info.NumElements` becomes a big number causing an  AccessViolationException.

Here is my new code:

```csharp
public static class DhcpsApi
{
    public static unsafe List<string> DhcpEnumSubnets(string serverAddress)
    {
        uint resumeHandle = 0;
        uint preferredMaximum = 65536;
        DHCP_IP_ARRAY* enumInfo = null;

        var result = PInvoke.DhcpEnumSubnets(serverAddress,
            ref resumeHandle,
            preferredMaximum,
            ref enumInfo,
            out uint elementsRead,
            out uint elementsTotal);

        if (result.IsEqual(DhcpErrorCode.ERROR_NO_MORE_ITEMS))
            throw new Exception();

        List<uint> nativeSubnets = new();
        List<string> subnets = new();

        while (result.IsSuccess())
        {
            try
            {
                for (uint i = 0; i < elementsRead; i++)
                {
                    DHCP_IP_ARRAY info = enumInfo[i];
                    for (uint j = 0; j < info.NumElements; j++)
                    {
                        uint element = info.Elements[j];
                        nativeSubnets.Add(element);
                    }
                }
                subnets = nativeSubnets.Select(s => DhcpHelper.NativeIpAddressToString(s)).ToList();
            }
            finally
            {
                DhcpRpcFreeMemory(enumInfo);
            }
        }
        return subnets;
    }

    [DllImport("dhcpsapi.dll", SetLastError = true)]
    static unsafe extern void DhcpRpcFreeMemory(void* BufferPointer);
}

public enum DhcpErrorCode : uint
{
    SUCCESS = 0,
    ERROR_NO_MORE_ITEMS = 259
}

public static class UintExtensions
{
    public static bool IsEqual(this uint a, DhcpErrorCode b)
    {
        if (a == (uint)b)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    public static bool IsSuccess(this uint a)
    {
        if (a.IsEqual(DhcpErrorCode.SUCCESS))
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}

public static class DhcpHelper
{
    public static string NativeIpAddressToString(uint ipAddress)
    {
        IPAddress ipAddressClass = new(ipAddress);
        string[] stringIpAddress = ipAddressClass.ToString().Split('.');
        return stringIpAddress[3] + "." + 
                   stringIpAddress[2] + "." + 
                   stringIpAddress[1] + "." + 
                   stringIpAddress[0];
    }

    public static uint StringIpAddressToNative(string ipAddress)
    {
        IPAddress ipAddressClass = IPAddress.Parse(ipAddress);
        byte[] ipAddressBytes = ipAddressClass.GetAddressBytes();
        uint nativeIpAddress = (uint)ipAddressBytes[0] << 24;
        nativeIpAddress += (uint)ipAddressBytes[1] << 16;
        nativeIpAddress += (uint)ipAddressBytes[2] << 8;
        nativeIpAddress += (uint)ipAddressBytes[3];
        return nativeIpAddress;
    }
}
```

Thank you in advance.

**Reply by AArnott** [reply]
I think your bug is that you don't actually call `DhcpEnumSubnets` within your loop. You call it once and assign `result`, then you loop over `result` forever because you never reassign it.
The reason for the crash is you released memory and then access it again on the next iteration of your loop. 

In my code example, I call the method *within* the loop, so with each iteration I'm getting new memory to read, reading it, and freeing it. Then as my loop condition I ask for more.

BTW, I updated my code snippet slightly because I think you have to actually free each individual array that is allocated throughout the structure to avoid a leak.

**Reply by stiezinger** [reply]
Thank you AArnott for the response,
I have changed the method call to this:

```csharp
public static unsafe List<string> DhcpEnumSubnets(string serverAddress)
{
    uint resumeHandle = 0;
    uint preferredMaximum = 65536;
    DHCP_IP_ARRAY* enumInfo = null;
    List<string> subnets = new();
    while (PInvoke.DhcpEnumSubnets(serverAddress,
        ref resumeHandle,
        preferredMaximum,
        ref enumInfo,
        out uint elementsRead,
        out uint elementsTotal).IsEqual(DhcpErrorCode.SUCCESS))
    {
        try
        {
            for (uint i = 0; i < elementsRead; i++)
            {
                DHCP_IP_ARRAY info = enumInfo[i];
                for (uint j = 0; j < info.NumElements; j++)
                {
                    uint nativeSubnetAddress = info.Elements[j];
                    var subnetAddress = DhcpHelper.NativeIpAddressToString(nativeSubnetAddress);
                    subnets.Add(subnetAddress);
                }
            }
        }
        finally
        {
            for (uint i = 0; i < elementsRead; i++)
            {
                DhcpRpcFreeMemory(enumInfo[i].Elements);
            }
            DhcpRpcFreeMemory(enumInfo);
        }
    }
    return subnets;
}
```

Unfortunettly it is still causing an infinite loop in the second iteration of the outer for loop.
By removing the outer loop and assigning `info` to the first index of `enumInfo` the method gives me the desired result.
I had to remove the loop in the finally block to call the method successfully.

```csharp
try
{
    DHCP_IP_ARRAY info = enumInfo[0];
    for (uint j = 0; j < info.NumElements; j++)
    {
        uint nativeSubnetAddress = info.Elements[j];
        var subnetAddress = DhcpHelper.NativeIpAddressToString(nativeSubnetAddress);
        subnets.Add(subnetAddress);
    }
}
finally
{
    DhcpRpcFreeMemory(enumInfo);
}
```

Maybe this is not the "right way". 
I dont know, for now it works, I will see if this works with other methods from the dhcpsapi.dll.
Do you think I can use the method like this?

**Reply by AArnott** [reply]
Without the outer loop, I think some DHCP configurations may not produce the full list. But I am not running a DHCP server so I can't test any of this locally to say for sure. I'm glad you have something that works for you.

**Reply by stiezinger** [reply]
Thank you
